Add DeclareAsNullable refactoring

jcouv opened the pull request:
The purpose for this refactoring is to assist in migrating a project to use the nullable feature.
When we see a comparison of a non-nullable with `null` (such as `nonNullable == null` or `nonNullable != null`), there is a good chance that `nonNullable` should in fact be declared as nullable instead. This refactoring helps do that with one click.
Note this is only a good chance, because such null checks could also be defensive programming, especially when it comes to public APIs (you declare them as non-nullable, but check anyways).

![image](https://user-images.githubusercontent.com/12466233/39670718-8077c4d0-50bf-11e8-8400-879f1d3cd328.png)

From discussion with Chuck, I will switch this to an analyzer/fixer so that it's more discoverable. But I won't offer a FixAll.

This PR was recreated from https://github.com/dotnet/roslyn/pull/26661 after the target branch was closed, so the first round of comments are in the original PR (sorry about that)

CyrusNajmabadi commented:
Let me know when you'd lke eyes :)

jcouv commented:
@CyrusNajmabadi Will do. I still have to address your feedback from the original PR (https://github.com/dotnet/roslyn/pull/26661).
If you have time, might you be able to look at https://github.com/dotnet/roslyn/pull/31843 (SignatureHelp on incomplete invocation)? I think I had addressed your initial feedback on factoring more code. Thanks

CyrusNajmabadi commented:
> If you have time, might you be able to look at #31843 (SignatureHelp on incomplete invocation)? I think I had addressed your initial feedback on factoring more code. Thanks

Sure!

jasonmalinowski commented:
@jcouv Is this something you need us to pick up?

jcouv commented:
@jasonmalinowski I'll revive this PR in the coming week.

jasonmalinowski commented:
@jcouv How is it coming? :smile:

jcouv commented:
@jasonmalinowski I refreshed this PR (rebased on latest master, converted refactoring to an analyzer+fixer, addressed some feedback from original PR). I'll take a look at handle a few more simple scenarios (`x is null`).

jcouv commented:
@jasonmalinowski I've addressed a couple more simple/common scenarios. We can add more later on.
Please take a look when you have some time.
I'm not sure if this should go into `master` or some other branch. Let me know.

sharwell commented:
> But I won't offer a FixAll.

It's fine to offer a Fix All. These are often easy to handle during constrained reviews, such as Fix All in Document. Most cases where we removed a Fix All because users might be bitten we ended up adding it back later because it was hurting cases where it was actually useful.

jcouv commented:
Rebased and resolved conflicts to revive this PR.

jcouv commented:
Added support for a few cases and added FixAll. Here's the result:
![image](https://user-images.githubusercontent.com/12466233/71790043-3adf0f00-2fe3-11ea-9d61-d004cef6be43.png)

As a reminder, the difference between DeclareAsNullable and PossiblyDeclareAsNullable is that the former involves high-confidence (assignment of `null` literal involved) whereas the latter is lower confidence (null test).

jcouv commented:
@jasonmalinowski @dotnet/roslyn-ide for review. Thanks

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/Diagnostics/Nullable/CSharpPossiblyDeclareAsNullableCodeFixTests.cs:
```diff
@@ -0,0 +1,628 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
"possibly declare as nullable" is an odd name...

jcouv replied:
[Copied from elsewhere] Here are my thoughts on tooling for nullability:

I see three categories of situations for making a type nullable:
1) assignments of literal null (for example, `return null;` offers to fix the return type. `DeclareAsNullable` fixer already shipped in VS)
2) null tests (present PR), such as `x is null` or `x?.Method()`.
3) assignments of maybe-null values (not implemented). For example, `return maybeNull;` would offer to fix the return type. This is similar to [1] but should be separate in my opinion, because I would likely want to separate the FixAll for [1] from that for [3].

Later on, we may want to also discuss fixers to adjust an override to match the overridden definition, or vice-versa.

----

In that context, we already have [1] (DeclareAsNullable), and this PR implements a chunk of [2]. I'm not sure what to call it. The main point is that the likelihood that you want this fix is lower than that of `DeclareAsNullable` which is high-confidence ;-)

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableCodeFixProvider.cs:
```diff
@@ -0,0 +1,140 @@
+        private async Task<TypeSyntax> TryGetFixableTypeAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true);
```
`getInnermostNodeForTie` is very likely correct.  however, it's worthwihle to add a test to make sure we don't regress that.  the easiest is to take one of your use cases and put it in the argument of some invocatoin call.  This will ensure that we get the expression node and not the argument node.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableCodeFixProvider.cs:
```diff
@@ -0,0 +1,140 @@
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true);
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var symbolToFix = CSharpPossiblyDeclareAsNullableDiagnosticAnalyzer.IsFixable(node, model);
```
haveing `IsFixable` return a `SymbolToFix` seems strange... i would expect GetSymbolToFix or TryGetSymbolToFix.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableCodeFixProvider.cs:
```diff
@@ -0,0 +1,140 @@
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var symbolToFix = CSharpPossiblyDeclareAsNullableDiagnosticAnalyzer.IsFixable(node, model);
+
+            var declarationLocation = symbolToFix.Locations[0];
```
are we guaranteed to always have at least one location.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableCodeFixProvider.cs:
```diff
@@ -0,0 +1,140 @@
+                return null;
+            }
+
+            return typeToFix;
```
the null check wasn't necessary it seems.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableCodeFixProvider.cs:
```diff
@@ -0,0 +1,140 @@
+    /// For example: `nonNull == null`, `nonNull?.Property`
+    /// </summary>
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.PossiblyDeclareAsNullable), Shared]
+    internal class CSharpPossiblyDeclareAsNullableCodeFixProvider : CodeFixProvider
```
i'm surprised this is nto a SyntaxEditorBasedCodeFixProvider.  Those are usually easier and safer to provide FixAll support for.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableCodeFixProvider.cs:
```diff
@@ -0,0 +1,140 @@
+            var typeNode = declarationLocation.FindNode(getInnermostNodeForTie: true, cancellationToken);
+
+            var typeToFix = TryGetTypeToFix(typeNode);
+            if (typeToFix == null || typeToFix is NullableTypeSyntax)
```
all callers seem to need to check for NullableTypeSyntax.  Consider pushing that check inside so that the caller doesn't have to care.

CyrusNajmabadi replied:
also, thi logic looks duplicated with TryGetFixableTypeAsync, can we not reuse that?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableCodeFixProvider.cs:
```diff
@@ -0,0 +1,140 @@
+        {
+            var node = diagnostic.Location.FindNode(getInnermostNodeForTie: true, cancellationToken);
+
+            var model = await document.GetSemanticModelAsync().ConfigureAwait(false);
```
missing cancellationToken

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableCodeFixProvider.cs:
```diff
@@ -0,0 +1,140 @@
+            }
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var fixedType = SyntaxFactory.NullableType(typeToFix.WithoutTrivia()).WithTriviaFrom(typeToFix);
```
this looks right.  but do we have trivia tests?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableCodeFixProvider.cs:
```diff
@@ -0,0 +1,140 @@
+                    }
+
+                    return method.ReturnType;
+            }
```
what about foreach cases?  aren't the declarations special there?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableCodeFixProvider.cs:
```diff
@@ -0,0 +1,140 @@
+                    return parameter.Type;
+
+                case VariableDeclaratorSyntax declarator:
+                    if (declarator.IsParentKind(SyntaxKind.VariableDeclaration))
```
`if (declarator.Parent.IsKind(SyntaxKind.variablDeclaration, out VariableDeclaration declaration)) ...`

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableCodeFixProvider.cs:
```diff
@@ -0,0 +1,140 @@
+            => ImmutableArray.Create(IDEDiagnosticIds.PossiblyDeclareAsNullableDiagnosticId);
+
+        // No support for FixAll at the moment. For example, in public API code this fix is likely incorrect.
+        public override FixAllProvider GetFixAllProvider() => null;
```
i'm a bit confused by this.  i thought you said fix-all was supported. :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,108 @@
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpPossiblyDeclareAsNullableDiagnosticAnalyzer :
+        DiagnosticAnalyzer
```
Derive from AbstractCodeStyleDiagnosticAnalyzer or AbstractBuiltInCodeStyleDiagnosticAnalyzer and you can cut out a lot of this boilerplate.  taht's what most IDE analyzers do.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,108 @@
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
```
for example, you shouldnt' have to do this.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,108 @@
+                IParameterSymbol parameter => parameter.Type,
+                ILocalSymbol local => local.Type,
+                IPropertySymbol property => property.Type,
+                IMethodSymbol method when method.IsDefinition => method.ReturnType,
```
can you comment why you have this `when` clause?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
```
`#nullable enable` on all these new files.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,108 @@
+                }
+            }
+
+            if (node.IsKind(SyntaxKind.IsPatternExpression))
```
`.IsKind(..., out IsPatternExpression patternExpr)`. (and the rest of the checks)

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpPossiblyDeclareAsNullableDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,108 @@
+            var fixableSymbol = IsFixable(context.Node, context.SemanticModel);
+            if (fixableSymbol != null)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(s_possiblyDeclareAsNullableRule, context.Node.GetLocation(), fixableSymbol.Name));
```
so, if i understand this correctly, all these variables will have `...` under them, but there is no way for the user to enable/disable this?  Is that the right experience we want here?

CyrusNajmabadi commented:
Slightly concerning to me that hte preview here shows an error in the resultant code:

![image](https://user-images.githubusercontent.com/4564579/72179812-9706b200-339a-11ea-985a-a64f0780f21b.png)

CyrusNajmabadi commented:
Done with first pass.  Happy to help this along.  Lots of institutional IDE knowledge in this area :)

jcouv commented:
> Slightly concerning to me that hte preview here shows an error in the resultant code

How did this occur? Is this a shipped preview, or a custom preview with this PR?

With the fix that shipped (`DeclareAsNullable`), if your code has `#nullable enable`, then the fix should not produce an error, and if it doesn't have `#nullable enable`, then the fix should not be offered. 
I validate that behavior just now (on 16.5p1).

jcouv commented:
In our experience with migrating code to add nullability annotations, this hasn't been much of a pain point.
I'll go ahead and close this PR and keep the branch so it could be revived if deemed useful.
